core/src/main/scala/org/http4s/rho/AuthedRhoService.scala

```diff
@@ -0,0 +1,65 @@
+package org.http4s
+package rho
+
+import org.http4s.rho.bits.PathAST.TypedPath
+import org.http4s.AuthedService
+import org.log4s.getLogger
+import shapeless.{HNil, HList}
+
+/** Constructor class for defining routes
+  *
+  * The [[AuthedRhoService]] provides a convenient way to define routes in a style
+  * similar to scalatra etc by providing implicit conversions and an implicit
+  * [[CompileService]] inside the constructor.
+  *
+  * {{{
+  *   val srvc = new AuthedRhoService[U] {
+  *     POST / "foo" / pathVar[Int] +? param[String]("param") |>> { (req: Request, p1: Int, param: String) =>
+  *       val authInfo: U = getAuth(req)
+  *       ...
+  *     }
+  *   }
+  *
+  * }}}
+  *
+  * @tparam U authInfo type for this service.
+  * @param routes Routes to prepend before elements in the constructor.
+  *
+  */
+class AuthedRhoService[U](routes: Seq[RhoRoute[_ <: HList]] = Vector.empty)
```

Comment:
What I mean is that function composition gives you the vast majority of stuff you need. For example, the class below would work for you:
```scala
class AuthContext[T] {
  private val key = AttributeKey[U]("key")
  def getUser(req: Request): T = req.attributes.get(key).get
  def transform(rhoService: Service[Request, Response]): Service[AuthedRequest[U], Response] =
    Service.lift { case AuthedRequest(authInfo, req) =>
      rhoService(req.withAttribute[U](key, authInfo))
    }
}
```

Then you just need to have this instance's `getUser` method when you construct your `RhoService` so you can actually get your user, and then use the same instance to transform it into an `AuthedService`. Note that this isn't a rho specific tool. The only downside is that nobody is stopping you from forgetting the second part and then `getUser` will explode since it clearly won't find the user in the attribute map. That last part is the only thing that `AuthedRhoService` really buys us.